### PR TITLE
Add txScrambler.

### PR DESCRIPTION
**Pull request by cabrador**
This PR adds func that sorts transactions order  using `XOR` on transaction hash. Furhtermore, if there are transactions from same sender, these transactions are sorted by the nonce.

### CODE REVIEW

File: gossip/tx_scrambler.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+	hash   common.Hash
+	sender common.Address
+	nonce  uint64
+}
+
+type scramblerSortFunc func(entries []*scramblerEntry, seed []byte) []*scramblerEntry
+
+// getExecutionOrder first removes any entries with duplicate hashes, then sorts the list by XORed hashes.
+// Furthermore, if there are entries with same sender, these entries are sorted by their nonce (lower comes first).
+func getExecutionOrder(entries []*scramblerEntry, sortFunc scramblerSortFunc) []*scramblerEntry {
```
Comment (HerbertJordan) [inline gossip/tx_scrambler.go:21]: This implementation does not follow the specification in the doc. In particular, it is not deterministic. This test fails:
```
func TestGetExecutionOrder_SortIsDeterministic2(t *testing.T) {
	entryA := &scramblerEntry{
		hash: common.Hash{1},
	}
	entryB := &scramblerEntry{
		hash: common.Hash{2},
	}

	res1 := getExecutionOrder([]*scramblerEntry{entryA, entryB}, builtInSort)
	res2 := getExecutionOrder([]*scramblerEntry{entryB, entryA}, builtInSort)

	if !reflect.DeepEqual(res1, res2) {
		t.Fatal("slices have different order - algorithm is not deterministic")
	}
}
```

File: gossip/tx_scrambler.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+		}
+
+		// if we have not found all indexes of given address, append and continue
+		if len(idxs) != occurrence {
+			indexMap[entry.sender] = append(idxs, i)
+			continue
+		}
+	}
+
+	for _, idxs := range indexMap {
```
Comment (HerbertJordan) [inline gossip/tx_scrambler.go:82]: This is an `O(N^2)` algorithm -- we should try to limit it to `O(NlogN)`.

**Review by HerbertJordan — CHANGES_REQUESTED** [review]

File: gossip/tx_scrambler_test.go [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+	)
+	size := int64(10)
+
+	for _ = range numLoops {
+		b.Run(fmt.Sprintf("builtIn_%d", size), func(b *testing.B) {
+			for i := 1; i <= b.N; i++ {
+				b.StopTimer()
+				entries := createRandomScramblerTestInput()
+				b.StartTimer()
+				getExecutionOrder(entries)
```
Comment (HerbertJordan) [inline gossip/tx_scrambler_test.go:130]: The `size` is nowhere used in this benchmark ... it should be odd that the benchmark probably reports the same time for any size of input.

**Review by HerbertJordan — CHANGES_REQUESTED** [review]
The algorithm looks good, but it should be extended to handle situations where the sender and nonce are equivalent, but the hash is different. Right now, we can make the tie-breaker the hash in this case, but in a follow-up we may want to use the gas price.

Besides that, I added a few comments on the testing and benchmarking. I think this needs another round of improvements.

File: gossip/tx_scrambler.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+	slices.SortFunc(list, func(a, b *scramblerEntry) int {
+		aX = xorBytes32(a.hash, salt)
+		bX = xorBytes32(b.hash, salt)
+		return bytes.Compare(aX[:], bX[:])
+	})
+}
+
+// unifyEntries removes any transactions with duplicate hashes and creates the XOR salt from the unique tx list.
+// Furthermore, it returns whether given list of entries contains duplicate addresses.
+func unifyEntries(entries []*scramblerEntry) ([]*scramblerEntry, [32]byte, bool) {
```
[the remark this answers is not in the document]
Comment (cabrador) [reply gossip/tx_scrambler.go:74]: I didnt want to to split it since it can be done in one loop. I honestly like `analyseEntryList`

File: gossip/tx_scrambler_test.go [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+
+	entries = getExecutionOrder(entries)
+
+	if got, want := len(entries), size/2; got != want {
+		t.Errorf("duplicate entries does not seem to be removed, got size: %v, want size: %v", got, want)
+	}
+}
+
+func TestTxScrambler_GetExecutionOrder_SortsSameSenderByNonce(t *testing.T) {
+	entries := createScramblerTestInputWithRepeatedAddresses()
```
[the remark this answers is not in the document]
Comment (cabrador) [reply gossip/tx_scrambler_test.go:103]: Added loop which increments number of repeated senders and the total number of entries.

File: gossip/tx_scrambler_test.go [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+			if entries[i].sender == entries[j].sender {
+				if entries[i].nonce > entries[j].nonce {
+					t.Errorf("incorrect nonce order %d must be before %d", entries[j].nonce, entries[i].nonce)
+				}
+			}
+		}
+	}
+}
+
+func TestTxScrambler_GetExecutionOrder_SortIsDeterministic(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (cabrador) [reply gossip/tx_scrambler_test.go:77]: Added two different tests. One for duplicate data and one for repeated data.
Both tests run in loop with increasing number of entries

File: gossip/tx_scrambler.go [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+	// no need to sort more
+	if !hasDuplicateAddresses {
+		return uniqueList
+	}
+
+	return sortTransactionsByNonce(uniqueList)
+}
+
+// sortTransactionsByNonce finds any duplicate senders and sorts their transactions by nonce ascending.
+func sortTransactionsByNonce(entries []*scramblerEntry) []*scramblerEntry {
```
Comment (HerbertJordan) [inline gossip/tx_scrambler.go:30]: This operation is not deterministic. See for example this test case:
```

func TestTxScrambler_SortTransactionsByNonce_IsDeterministic(t *testing.T) {
	entries := []*scramblerEntry{}
	for i := range 10 {
		entries = append(entries, &scramblerEntry{
			hash: common.Hash{byte(i)},
		})
	}

	printOrder := func(entries []*scramblerEntry) string {
		builder := strings.Builder{}
		for _, entry := range entries {
			builder.WriteString(fmt.Sprintf("%d ", entry.hash[0]))
		}
		return builder.String()
	}

	want := sortTransactionsByNonce(entries)
	for range 10 {
		input := slices.Clone(want)
		shuffleEntries(input)
		got := sortTransactionsByNonce(input)
		if !reflect.DeepEqual(want, got) {
			t.Errorf("sort is not deterministic, want %v, got %v",
				printOrder(want),
				printOrder(got),
			)
		}
	}
}
```

**Review by HerbertJordan — CHANGES_REQUESTED** [review]
The algorithm you have is not (yet) deterministic -- and although there are tests for it, they failed to detect this.

Please re-evaluate your testing strategy and try to make it ironclad.

File: gossip/tx_scrambler.go (hunk 7 above)
Comment (HerbertJordan) [reply gossip/tx_scrambler.go:30]: Actually, forget that. I was writing this up too fast, I have to rethink this test case.

File: gossip/tx_scrambler_test.go [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+		}
+		seenHashes[entry.hash] = true
+	}
+}
+
+func createRandomSalt() [32]byte {
+	var salt = [32]byte{}
+	for i := 0; i < 32; i++ {
+		salt[i] = byte(rand.Intn(math.MaxUint8))
+	}
```
Comment (HerbertJordan) [inline gossip/tx_scrambler_test.go:340]: Simpler:
```suggestion
	rand.Read(salt[:])
```

File: gossip/tx_scrambler_test.go [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+	"github.com/ethereum/go-ethereum/common"
+	"golang.org/x/exp/rand"
+	"math"
+	"math/big"
+	"reflect"
+	"testing"
+)
+
+func TestTxScrambler_AnalyseEntryList_RemovesDuplicateTransactions(t *testing.T) {
+	entries := createScramblerTestInputRepeatedHashes(2, 10)
```
Comment (HerbertJordan) [inline gossip/tx_scrambler_test.go:14]: With these test input generation functions being out-factored and parameterized it is very difficult to understand what is actually tested in these functions. 

I would argue that in unit tests clarity is favorable over code re-use. It should be possible for anyone by just reading the code in the test to figure out what the test is about and whether the test itself is doing what the label promises.

A test like this would be much easier to comprehend:
```
func TestTxScrambler_AnalyseEntryList_RemovesDuplicateTransactions2(t *testing.T) {
	entries := []*scramblerEntry{
		{hash: common.Hash{2}},
		{hash: common.Hash{1}},
		{hash: common.Hash{1}},
	}
	entries, _, _ = analyseEntryList(entries)

	want := []*scramblerEntry{
		{hash: common.Hash{2}},
		{hash: common.Hash{1}},
	}
	if !reflect.DeepEqual(want, entries) {
		t.Errorf("want %v, got %v", want, entries)
	}
}
```

Or it might be even extended to multiple different input cases without harming readability a lot:
```
func TestTxScrambler_AnalyseEntryList_RemovesDuplicateTransactions3(t *testing.T) {
	examples := [][]*scramblerEntry{
		{},
		{
			{hash: common.Hash{1}},
		},
		{
			{hash: common.Hash{1}},
			{hash: common.Hash{1}},
		},
		{
			{hash: common.Hash{1}},
			{hash: common.Hash{1}},
			{hash: common.Hash{1}},
		},
		{
			{hash: common.Hash{1}},
			{hash: common.Hash{2}},
			{hash: common.Hash{1}},
		},
		{
			{hash: common.Hash{1}},
			{hash: common.Hash{2}},
			{hash: common.Hash{3}},
			{hash: common.Hash{2}},
			{hash: common.Hash{1}},
		},
	}

	for _, example := range examples {
		t.Run("", func(t *testing.T) {
			result, _, _ := analyseEntryList(example)

			seen := map[common.Hash]struct{}{}
			for _, entry := range result {
				if _, seen := seen[entry.hash]; seen {
					t.Fatalf("duplicate hash %v", entry.hash)
				}
				seen[entry.hash] = struct{}{}
			}
		})
	}
}
```

File: gossip/tx_scrambler_test.go [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+	entries, _, _ = analyseEntryList(entries)
+
+	checkDuplicateHashes(t, entries)
```
Comment (HerbertJordan) [inline gossip/tx_scrambler_test.go:17]: nit: This test would pass if the implementation of `analyseEntryList` would just always return an empty list.

File: gossip/tx_scrambler_test.go [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+	res1 := createScramblerTestInputRepeatedAddr(2, 10)
+	res2 := deepCopyEntries(res1)
+	// shuffle one array
+	shuffleEntries(res2)
+
+	for i := 0; i < 10; i++ {
+		salt := createRandomSalt()
+		scrambleTransactions(res1, salt)
+		for j := 0; j < 10; j++ {
+			scrambleTransactions(res2, salt)
```
Comment (HerbertJordan) [inline gossip/tx_scrambler_test.go:57]: I think `res2` should be shuffled each time before calling `scrambleTransactions`

File: gossip/tx_scrambler_test.go [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+		b.Run(fmt.Sprintf("builtIn_%d", size), func(b *testing.B) {
+			for i := 1; i <= b.N; i++ {
+				b.StopTimer()
+				entries := createRandomScramblerTestInput(size)
+				b.StartTimer()
+				getExecutionOrder(entries)
+			}
+		})
+		size = size * multiplier
+	}
```
Comment (HerbertJordan) [inline gossip/tx_scrambler_test.go:204]: simpler:
```suggestion
	for size := uint64(10); size < 100_000; size *= 10 {
		b.Run(fmt.Sprintf("size_%d", size), func(b *testing.B) {
			for i := 1; i <= b.N; i++ {
				b.StopTimer()
				entries := createRandomScramblerTestInput(size)
				b.StartTimer()
				getExecutionOrder(entries)
			}
		})
	}
```

**Review by HerbertJordan — CHANGES_REQUESTED** [review]
I think the algorithm is good, with the exception of how the same nonces are handled. This should be improved.

About the tests: it takes a lot of time to think them through due to the extra nesting of generator functions. The goal for tests is different than for regular code. In tests it should be plane obvious to the reader of the code what is going on, without knowing a lot of the other code around it. If several functions need to be studied in detail to understand a test, the test is too complicated.

The reasoning is that tests themself are not tested. Reviews are the last line of defense. You need to write them in a way that allows anybody to gain trust in the fact that these tests are doing what they are supposed to do.

Please give those tests another overhaul with the aim of making it easy for reviewers to understand what is going on.

File: gossip/tx_scrambler.go [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+
+func xorBytes32(a, b [32]byte) (dst [32]byte) {
+	for i := 0; i < 32; i++ {
+		dst[i] = a[i] ^ b[i]
+	}
+	return
+}
+
+// deepCopyEntries returns deep copy of entries.
+func deepCopyEntries(entries []*scramblerEntry) []*scramblerEntry {
```
Comment (HerbertJordan) [inline gossip/tx_scrambler.go:120]: nit: only needed in tests => move to test file

File: gossip/tx_scrambler.go [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+			return -1
+		}
+		// if nonce is equal, sort by gas
+		if a.gas > b.gas {
+			return -1
+		} else {
+			// if both nonce and gas is equal, order between those do not matter
+			return 1
+		}
+	})
```
Comment (HerbertJordan) [inline gossip/tx_scrambler.go:52]: To guarantee a total order, you need to consider the case where `a.sender == b.sender && a.nonce == b.nonce && a.gas == b.gas`; in this case, you should use the transaction hash as a tie breaker.

File: gossip/tx_scrambler.go [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+			return 1
+		}
+		if a.nonce < b.nonce {
+			return -1
+		}
+		// if nonce is equal, sort by gas
+		if a.gas > b.gas {
+			return -1
+		} else {
+			// if both nonce and gas is equal, order between those do not matter
```
Comment (HerbertJordan) [inline gossip/tx_scrambler.go:49]: Attention: It does matter!

The sort function expects that `Cmp(a,b)==1` if and only if `Cmp(b,a)==-1`; If you have two equal elements `a` and `b` right now, you would return `1` in both cases.

File: gossip/tx_scrambler_test.go [hunk 16]
```diff
@@ -0,0 +1,597 @@
...
+		{hash: common.Hash{3}},
+		{hash: common.Hash{2}},
+		{hash: common.Hash{1}},
+	}
+
+	shuffleEntries(entries)
+	result, _, _ := analyseEntryList(entries)
+	if len(result) == 0 {
+		t.Fatal("analyseEntryList returned empty list")
+	}
```
Comment (HerbertJordan) [inline gossip/tx_scrambler_test.go:25]: This can be stricter:
```suggestion
	if len(result) != 3 {
		t.Fatalf("unexpected length of result list, wanted 3, got %d", len(result))
	}
```

File: gossip/tx_scrambler_test.go [hunk 17]
```diff
@@ -0,0 +1,597 @@
...
+			nonce:  2,
+		},
+		{
+			hash:   common.Hash{5},
+			sender: common.Address{1},
+			nonce:  2,
+		},
+	}
+
+	sortTransactionsWithSameSender(entries)
```
Comment (HerbertJordan) [inline gossip/tx_scrambler_test.go:146]: The input is already sorted in the desired way. So if this function does nothing, the test passes.

File: gossip/tx_scrambler_test.go [hunk 18]
```diff
@@ -0,0 +1,597 @@
...
+	}
+
+	shuffleEntries(entries)
+	sortTransactionsWithSameSender(entries)
+	for i := 0; i < len(entries); i++ {
+		for j := i + 1; j < len(entries); j++ {
+			if entries[i].sender == entries[j].sender {
+				if entries[i].nonce > entries[j].nonce {
+					t.Errorf("incorrect nonce order %d must be before %d", entries[j].nonce, entries[i].nonce)
+				}
```
Comment (HerbertJordan) [inline gossip/tx_scrambler_test.go:193]: nit: seems to be irrelevant since all nonces are 1

File: gossip/tx_scrambler_test.go [hunk 19]
```diff
@@ -0,0 +1,597 @@
...
+					hash:   common.Hash{3},
+					sender: common.Address{3},
+					nonce:  3,
+					gas:    3,
+				},
+				{
+					hash:   common.Hash{4},
+					sender: common.Address{2},
+					nonce:  2,
+					gas:    4,
```
Comment (HerbertJordan) [inline gossip/tx_scrambler_test.go:441]: If I understand the test-case setup correctly, this should be a 2 and in the entry below a 1

**Review by HerbertJordan — CHANGES_REQUESTED** [review]
Looks good, but the gas-tie breaker still needs to be added.

File: gossip/tx_scrambler_test.go (hunk 17 above)
Comment (cabrador) [reply gossip/tx_scrambler_test.go:146]: Added shuffle before calling `sortTransactionsWithSameSender`

File: gossip/tx_scrambler_test.go (hunk 19 above)
Comment (cabrador) [reply gossip/tx_scrambler_test.go:441]: You are right, thanks

File: gossip/tx_scrambler.go [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+	"github.com/ethereum/go-ethereum/common"
+	"slices"
+)
+
+// scramblerEntry stores meta information about transaction for sorting and filtering them.
+type scramblerEntry struct {
+	hash   common.Hash    // transaction hash
+	sender common.Address // sender of the transaction
+	nonce  uint64         // transaction nonce
+	gas    uint64         // transaction gas
```
Comment (HerbertJordan) [inline gossip/tx_scrambler.go:15]: `gas` => `gasPrice`

File: gossip/tx_scrambler_test.go [hunk 21]
```diff
@@ -0,0 +1,645 @@
...
+
+func TestTxScrambler_GetExecutionOrder_RandomInput(t *testing.T) {
+	// this tests these input sizes:
+	// 1, 4, 16, 64, 256, 1024
+	for i := int64(1); i <= 1024; i = i * 4 {
+		input := createRandomScramblerTestInput(i)
+		cpy := deepCopyEntries(input)
+		shuffleEntries(cpy)
+		input = filterAndOrderTransactions(input)
+		cpy = filterAndOrderTransactions(input)
```
Comment (HerbertJordan) [inline gossip/tx_scrambler_test.go:575]: Should this get `cpy` as an argument?
```suggestion
		cpy = filterAndOrderTransactions(cpy)
```

File: gossip/tx_scrambler_test.go [hunk 22]
```diff
@@ -0,0 +1,342 @@
...
+			t.Fatal("found duplicate hash in entries")
+		}
+		seenHashes[entry.hash] = true
+	}
+}
+
+func createRandomSalt() [32]byte {
+	var salt = [32]byte{}
+	for i := 0; i < 32; i++ {
+		salt[i] = byte(rand.Intn(math.MaxUint8))
```
[the remark this answers is not in the document]
Comment (HerbertJordan) [reply gossip/tx_scrambler_test.go:631]: This problem still seems to exist.

File: gossip/tx_scrambler_test.go (hunk 8 above)
Comment (HerbertJordan) [reply gossip/tx_scrambler_test.go:632]: This one as well ...

File: gossip/tx_scrambler_test.go [hunk 23]
```diff
@@ -0,0 +1,645 @@
...
+	}
+	return salt
+}
+
+// deepCopyEntries returns deep copy of entries.
+func deepCopyEntries(entries []*scramblerEntry) []*scramblerEntry {
```
Comment (HerbertJordan) [inline gossip/tx_scrambler_test.go:637]: nit: I think nowhere in these tests you actually need a deep copy and a deep equality since all entries are all constant. Shallow copies of slices and comparisons would do the job.

File: gossip/tx_scrambler.go [hunk 24]
```diff
@@ -0,0 +1,122 @@
...
+	return
+}
+
+func compareAsc[T cmp.Ordered](a T, b T) int {
+	return cmp.Compare(a, b)
+}
+
+func compareDesc[T cmp.Ordered](a T, b T) int {
+	return -cmp.Compare(a, b)
+}
```
Comment (HerbertJordan) [inline gossip/tx_scrambler.go:122]: nit: these functions only add minor benefits over just calling
```
cmp.Compare(a.nonce, b.nonce) // ascending order
```
or
```
cmp.Compare(b.nonce, a.nonce)  // descending order
```
where it is needed. Also, they are both only used once, so they there is no big utility in defining extra functions for this.

**Review by HerbertJordan — APPROVED** [review]
Looks good, only 1-2 minor issues and style comments. Then I think it is ready to merge.

File: gossip/tx_scrambler_test.go (hunk 8 above)
Comment (cabrador) [reply gossip/tx_scrambler_test.go:632]: Sorry I thought I factored this in.

File: gossip/tx_scrambler_test.go [hunk 25]
```diff
@@ -0,0 +1,680 @@
...
+
+// dummyScramblerEntry represents scramblery entry data used for testing
+type dummyScramblerEntry struct {
+	hash     common.Hash    // transaction hash
+	sender   common.Address // sender of the transaction
+	nonce    uint64         // transaction nonce
+	gasPrice uint64         // transaction gasPrice
+}
+
+func (s *dummyScramblerEntry) Unwrap() *types.Transaction {
```
Comment (HerbertJordan) [inline gossip/tx_scrambler_test.go:662]: What is this needed for?

**Review by HerbertJordan — APPROVED** [review]

File: gossip/tx_scrambler_test.go (hunk 25 above)
Comment (cabrador) [reply gossip/tx_scrambler_test.go:662]: archaic code. Removed

**Review by HerbertJordan — APPROVED** [review]
